### DISCUSSION TITLE
Installs/releases

### DISCUSSION THREAD

**Discussion by wormstein** [EdNekebno/DGTCentaurMods#54] [General]
Hi, I will now focus all on building a release to have a way to easy deploy the mods on a new board. I have several ideas that I want to explore.
Let me bring a first release and I we'll take it from there, In the mean time, I am open for suggestion here,,,

Apart from soldering the pins, I want the rest to be plug and play.

**Comment by EdNekebno** [comment]
Sounds like a plan. I think you're probably much better in this sort of thing than me - I have no ideas. Other than I should probably start to learn how to use git/github/branches so that the master branch has a bit more stability.

It's a shame about the unusual header pins - but a fairly easy soldering job. Having the rest plug and play as you say would be amazing. The DGT Centaur is close to being the most advanced chess computer available!

**Comment by wormstein** [comment]
Getting there

```
pi@dev-pi:~/dev/build $ sudo dpkg -i DGTCentaur_master_armhf.deb
Selecting previously unselected package dgtcentaurmods.
(Reading database ... 63133 files and directories currently installed.)
Preparing to unpack DGTCentaur_master_armhf.deb ...
Unpacking dgtcentaurmods (0) ...
Setting up dgtcentaurmods (0) ...
```
I managed to automate the process of building a release from our repo, I am now starting the work on script to asses and setup hardware configuration,, setup services and whatever else comes after install.

**Reply by EdNekebno** [reply]
It's sorcery to me. But it's looking good. So exciting. Meanwhile I'm just messing about making the web interface export pgns :)

**Comment by EdNekebno** [comment]
Thinking about the setup services, whatever bit . I remembered that for things like lichess that use SSL APIs I needed to turn on ntp to get a fairly accurate time.

**Reply by wormstein** [reply]
Noted

**Comment by wormstein** [comment]
I will try to configure the install to prompt for lichess API. Is that now on centaur.ini? Where to write it?

**Reply by EdNekebno** [reply]
It's not yet. I will do it when I next work on lichess.py though. I need to adapt lichess.py, the way to enter it with the board pieces, and the web interface. It's not a big job though - so I think for your stuff you can probably assume it is in centaur.ini and write it to there.

**Reply by EdNekebno** [reply]
Oh...and we should probably do the trick you did with centaur...we can hide it on the menu if they haven't given a key.

**Reply by wormstein** [reply]
> Oh...and we should probably do the trick you did with centaur...we can hide it on the menu if they haven't given a key.

Good idea. I'll handle this.

**Comment by wormstein** [comment]
I just did a pull request for the build automation.

build.sh does all the process of creating a ready-to-install .deb package,
To build a release from master you do:
```
cd build
./build.sh clean      # not needed if no other builds we're done
./build.sh master
```
.... wait and enjoy!

Installing the package:
`sudo apt install <name>.deb`

Uninstalling:
`sudo apt remove <name>.deb`
The remove process needs some tuning so we make sure everything gets removed. It's work in some way though.

If you just do `./build.sh` it will go interactive and will ask you if this is a build from master or you go for a tag.

**Comment by EdNekebno** [comment]
This project is all things I'm unfamiliar with - python, git, and deb packages :)
From what I read this is fantastic though. If I understand it it builds the deb file along with setup for the services and pip3 + required packages?
If so - can we magic it with bluetooth setup too? Or is that not the done thing? Then you'd just get your raspbian image and install this?
And if so does it need me to find a new sdcard, use this and install from scratch and work out what the heck I did to get bluetooth working in the first place? :)

**Comment by wormstein** [comment]
Yes, the builf.sh is staging everything to build the deb package. control file and postinst files go into it but at the same time they are modified by the build.sh. control file contains references to any other needed packages (I should add python3-pip there...)

Then, on the install, postinst is executed. It will check for SPI and serial port and enable them in config.txt. Then it will enable services ffor startup and also pip3 install the requirements.txt (will know automatically where it is).
And the final step is to ask for Lichess API and write it to the config.

I will add option to reboot to finish all.

Didn't forget about BT, I just need to recreate the whole process and see what steps are needed on a clean RPi image.

**Comment by EdNekebno** [comment]
Genius.
I sense you will be holding me to my promise of creating a development branch soon :)

**Comment by wormstein** [comment]
Nah, as long what we push to master is tested and don't gets released, it's ok... But anyway we can decide at any time if master is stabke enough and worth a new version release.

Also, sneaky ssh console over ttyS0 needs to be disabled at setup :))

**Comment by wormstein** [comment]
Do you think we have a stable enough master to go tomorrow for a beta release?

I feel optimistic enough to run the deb on my board. Shouldn't break anything reallly.

**Reply by EdNekebno** [reply]
I think it's relatively stable. For any release I have two thoughts: firstly the eboard feature is important to some people, so I think we need to be at the point we have working Bluetooth or can describe how to do it. If it's needed  I could probably find an sd-card, install raspbian lite, install the deb, and then try and work/out remember what commands I typed :) That might take a day or two though. I'm not 100% confident in the lichess code, it's contributed by somebody adapting my original and apparently the next version is much more stable, but I haven't got it yet and I'm thinking about adapting it to gamemanager - but I ran out of time this weekend. By which I think I'm saying that after next weekend I think we'll be further along. Or we can just call it alpha and just do exactly what we have now.

**Reply by wormstein** [reply]
I'd say the same regarding the Bluetooth. I'll turn my focus on this and see what I can do in the next 1-2 days. We'll reconsider then.

**Reply by EdNekebno** [reply]
For bluetooth apparently we need to change:
/etc/bluetooth/pin.conf :  now reads:   * *
/etc/machine-info: PRETTY_HOSTNAME=PCS-REVII-081500

and I've upgraded the pair.py script a bit.

**Reply by wormstein** [reply]
So, you want me to put `/etc/machine-info: PRETTY_HOSTNAME=PCS-REVII-081500` in pin.cof?
Apparently on a new RPi it is completly missing. I will create it,

**Reply by EdNekebno** [reply]
Sorry...there should be two files. in
/etc/bluetooth/pin.conf
you put: * *
in /etc/machine-info
you put: /etc/machine-info: PRETTY_HOSTNAME=PCS-REVII-081500
The first one tells it to accept any pin number from anything
The second one pretends to be a particular chess board for the lights

**Comment by EdNekebno** [comment]
Oooh...another thought. Can the deb pull the binary for the modified stockfish from somewhere or are we better having menu.py do that?

**Reply by wormstein** [reply]
I will add this download during deb installetion because it is clear the board has internet connection (needs SSH access to install)

There will be another installation option with a ready made SD card image that the user will just load the SD card, put the RPi in and wait the magic to happen. Will decide then where to put the SF in this case.

**Reply by EdNekebno** [reply]
If it's the image I guess we can just put the binary on there as long as we put the sources somewhere and make it obvious to meet the license.

**Reply by wormstein** [reply]
I would keep a clean image and just put in the trigger for setup at first boot. This way we don't have to build a new image every time we release because the image will self deploy the new release.

For the SF. I will build a version wirh the path to db to `/home/pi/DGTCentaurMods/db` and find a place in the cloud to serve it for download. But, in order to let in the future the user to select the setup location, means SF needs to be recompiled on deployment (because path needs to be modified depending on user input). Is not high priority yet because I don't ask for the option to change setup location.

**Reply by EdNekebno** [reply]
We can always cheat a bit and just put the path in a text file in the same folder as stockfish (so engines), then have stockfish read that to get the location to the db. It'll slow stockfish down a very tiny bit but it's a chess engine...nobody would notice! (i agree the setup location isn't a priority though...people are just going to install it as an easy way to get everything working).

**Reply by wormstein** [reply]
For now, I cannot compile SF. It is the one from your fork. Any clues?
```
nnue/evaluate_nnue.cpp:156:80: warning: requested alignment 64 is larger than 8 [-Wattributes]
       TransformedFeatureType transformedFeatures[FeatureTransformer::BufferSize];
                                                                                ^
nnue/evaluate_nnue.cpp:157:55: warning: requested alignment 64 is larger than 8 [-Wattributes]
     alignas(alignment) char buffer[Network::BufferSize];
                                                       ^
nnue/evaluate_nnue.cpp: In function ‘Stockfish::Eval::NNUE::NnueEvalTrace Stockfish::Eval::NNUE::trace_evaluate(const Stockfish::Position&)’:
nnue/evaluate_nnue.cpp:205:80: warning: requested alignment 64 is larger than 8 [-Wattributes]
       TransformedFeatureType transformedFeatures[FeatureTransformer::BufferSize];
                                                                                ^
nnue/evaluate_nnue.cpp:206:55: warning: requested alignment 64 is larger than 8 [-Wattributes]
     alignas(alignment) char buffer[Network::BufferSize];
                                                       ^
make[1]: *** No rule to make target 'half_ka_v2.o', needed by 'stockfish'.  Stop.
```

**Reply by EdNekebno** [reply]
Yup...from my fork. Is because I built NNUE into it. I can't imagine why - but maybe you need to do: make net    first?
or even:  make clean     and then    make net .

Although I only changed the Makefile to link in sqlite and uci.cpp so just uploaded those files to the original. I can give it a try on mine...maybe I missed something

**Reply by EdNekebno** [reply]
Okay...it's my mistake. I messed up the Makefile. New one coming....

**Reply by EdNekebno** [reply]
New makefile should do it.
If you don't have them you might need sqlite3 development installed for the linking.

**Reply by wormstein** [reply]
That worked. I thought it was my Pi because I was using my Pi4 to speed up things...
Thanks.

**Reply by EdNekebno** [reply]
If it's a choice between a technical problem and me being an idiot. Always assume it's me being an idiot :)

**Comment by wormstein** [comment]
Bluetooth setup done. Moving to see what I can do about Stockfish download...

**Comment by wormstein** [comment]
I will wipe tonight my dev Pi SD card and start from scratch and take a note of any additions I do to make sure a new user will be able to install it.

**Reply by EdNekebno** [reply]
Ooooh...exciting moment!

**Reply by wormstein** [reply]
Just did a test setup and there were no issues on installing. I only need to setup SF download tomorrow. That is uploaded to a google drive and shared.

Have a look in control file in build at "Depends:" that holds all required additional packages. Do you think any other is needed?
Do you know if pybluez was added with pip or apt? That will be the last one beeded...

**Reply by wormstein** [reply]
The very last test would be to place new imaged RPi inside the board so the software can start... I only have some time tomorrow evening or Wednesday all day.

**Comment by wormstein** [comment]
I'll put here: on a clean install, centaurmds-web.service fails to boot because it is looking for the dreaded epaper.jpg  :))

```
  File "/home/pi/DGTCentaurMods/web/app.py", line 143, in <module>
    moddate = os.stat(str(pathlib.Path(__file__).parent.resolve()) + "/../web/static/epaper.jpg")[8]
FileNotFoundError: [Errno 2] No such file or directory: '/home/pi/DGTCentaurMods/web/../web/static/epaper.jpg'
```
And there was another one I had to comment.

**Reply by EdNekebno** [reply]
I pushed a fix for this. Can you remember the other one?

**Comment by wormstein** [comment]
I will add an option to compile and integrate Stockfish in the build at the moment of creation. Sources will come from SF repo on my Github account, The default release will be without SF, it will be downloaded on install.

Now this makes me think of another thing. Let the installer check if centaur is installed, if not, change a setting in the ini file to point to other location of SF.

The default path in the ini file should be /home/pi/centaur/engines.

**Reply by EdNekebno** [reply]
Not entirely understanding. If there's no centaur then shouldn't we create that folder structure anyway and put the stockfish_pi in . Then people can play it directly with the uci option.

**Reply by wormstein** [reply]
That would do it too. Less code to deploy it.
